### ISSUE TITLE
[54] Dense-to-Sparse Gate for Mixture-of-Experts

### ISSUE THREAD

**Issue by dhkim0225** [dhkim0225/1day_1paper#82]
MoE 고르는 gate 바꿔서 효과적이였다는 논문
[paper](https://arxiv.org/pdf/2112.14397.pdf)

# Method

### Sparse Gate Network
기존 MoE 는 다음과 같은 방식으로 동작함. (Gshard)
![image](https://user-images.githubusercontent.com/16400591/147745184-b67c4803-d788-43ba-8341-2b7b9bce81b6.png)
W_g 는 DxN size 이고, N expert 의 weight를 의미함.

### Expert Network
그렇게 뽑은 softmax probability 를 이용해서 선택한 네트워크들끼리 weighted sum 을 해줬음.
e 는 expert network 임
![image](https://user-images.githubusercontent.com/16400591/147745926-7ce59d25-4100-432a-bc15-66279d9de762.png)

##  DENSE-TO-SPARSE GATE (DTS Gate)
처음에는 dense 하게 전부 routing 하다가 점점 expert 를 sparse 하게 가져가는 DTS Gate 제안
gumbel softmax 사용
![image](https://user-images.githubusercontent.com/16400591/147746849-5c3ccf90-11e5-44f6-b353-e436e3b13ef5.png)
ζ 는 Gumbel(0, 1) distribution에서 뽑은 extra noise
temperature τ 로 sparsity 조절 (처음에는 큰 temperature, 나중에는 작은 temperature 로 linear decay)
![image](https://user-images.githubusercontent.com/16400591/147746210-be5ebd07-89a3-488d-b34d-da3f05bd2794.png)

# Result
GPT-MoE 를 갖고 실험해 봤더니, 같은 perplexity 도달까지 2배 덜 iteration 돌면 됨.
1.42 배 빠르기까지 함.

![image](https://user-images.githubusercontent.com/16400591/147746425-6ca1e7f5-e64d-48be-be48-9bbe14511e49.png)
![image](https://user-images.githubusercontent.com/16400591/147746437-11fc6bab-cd79-40b6-9656-bc338b58e371.png)

추가로, FLOP 수에 따른 Validation PPL 비교.
더 많은 layer, 더 많은 expert 를 쓰는게 오히려 가성비가 좋다고 주장. 
(large scale은 아니라서, 꼭 정답인건 아니지만 꽤 재미있는 실험)
![image](https://user-images.githubusercontent.com/16400591/147746510-fd35b221-2311-42c2-935d-ecfdca7ee283.png)

## Ablation
temperature linear scheduling 에 따른 실험 결과
![image](https://user-images.githubusercontent.com/16400591/147747087-33828846-c825-4fa1-9148-2d5c8f0e3a20.png)

## Expert Balancing 관련 이슈.
몇몇 expert 만 학습이 치중되는 것을 막기 위해
1. GShard 가 했듯이  balance loss 를 사용할 수 있고
2. [Base-layers](http://proceedings.mlr.press/v139/lewis21a.html) paper 처럼 balanced assignment algorithm을 쓸 수도 있다.

Gshard 논문 방식으로 추가설명을 하자면, 
(1)은 training batch 의 총 token 수가 N 이라고 할 때, expert capacity 라는 것을 (N/E) 로 매번 정한다.
그리고, 정해진 expert capacity 를 넘게 token 들이 expert 들에 할당되면, 
해당 token 값들은 overflowed token 이라고 보고, next-layer에 그냥 skip-connection으로 넘긴다.
예를 들어 총 token 은 1000 개, expert 는 10개 가 있으면 expert 별 처리 token 은 100개까지만 할 수 있는 거고, 나머지는 skip-con으로 next-layer 에 넘겨버린다는 방식.

(2) 는 동일한 개수의 token을 모든 expert 에 통과시키는 방식이다.
약간은 무식한 방식을 사용하는데,
![image](https://user-images.githubusercontent.com/16400591/147748299-455eac4e-0dba-483a-a5ab-b6c5ebd50bc3.png)
`hungarian algorithm` 을 gpu parallel 하게 구현한, `auction algorithm` 을 사용해서 최대한 맞는 expert에 할당시켜주겠다는 아이디어이다.
즉, 토큰을 각 expert에 할당하기 위한 점수를 계산하고 auction algorithm 으로 expert 에 뿌려줌
이 점수를 `token-expert affinities` 라고 부르는데, 그냥 최종 logit 값 제일 커지게 어떻게든 permutation 을 구하는 것이다.
![image](https://user-images.githubusercontent.com/16400591/147749847-71b8d162-068f-4ea7-bdef-4ca10660c120.png)

요 방식의 full 코드는 fairseq 에 있다.
1. https://github.com/pytorch/fairseq/blob/fcca32258c8e8bcc9f9890bf4714fa2f96b6b3e1/fairseq/modules/base_layer.py
2. https://github.com/pytorch/fairseq/blob/fcca32258c/fairseq/clib/libbase/balanced_assignment.cpp

얘네는 balance loss 만 써봤는데 쓰나 안 쓰나 별 차이 없다고 한다.
balanced assignment algorithm 실험은 딱히 안 해봤다고 한다.
figure 의 (b)는 unbalance 에서 배워진 MoE weight 들인데, 충분히 잘 퍼져 있다고 주장한다.
![image](https://user-images.githubusercontent.com/16400591/147748487-c48a3aab-0d16-4dca-a71d-9dc055785eb0.png)

# 필자 의견
GShard 나 GLaM 과 같이 MoE 로 scaling 한 방식들에 대한 실험은 없어서 아쉽지만
MoE를 쓴다면 적용해봄직한 방법론
~balanced assignment algorithm 실험도 해주지... 궁금한데...~
